python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -0,0 +1,758 @@
+            ),
+        )
+
+        script = os.path.join(
+            os.path.dirname(__file__),
+            "scripts",
+            "execute_command.py"
+        )
+
+        output = None
+
+        try:
+            output = subprocess.check_output(
+                [sys.executable, script, args_file]
+            )
+        except subprocess.CalledProcessError:
+            if output:
+                self.logger.error(output)
+            raise
+        else:
+            self.logger.debug(output)
+
+        self.logger.debug("Command execution complete.")
+        self.host.reply(
+            dict(
+                retcode=self.COMMAND_SUCCEEDED,
+                out="Command executed successfully.",
+                err="",
+            ),
+        )
+
+    def get_actions(self, data):
+        """
+        RPC method that sends back a dictionary containing engine commands
+        for each pipeline configuration associated with the project.
+
+        :param dict data: The data passed down by the client. At a minimum,
+            the dict should contain the following keys: project_id, entity_id,
+            entity_type, pipeline_configs, and user.
+        """
+        # If we weren't sent a usable entity id, we can just query the first one
+        # from the project. This isn't a big deal for us, because we're only
+        # concerned with picking the correct environment when we bootstrap
+        # during a caching operation. This is going to be the situation when
+        # a page is pre-loading entity-level commands on page load, as opposed to
+        # passing down a specific entity that's been selected in an already-loaded
+        # page.
+        if "entity_id" in data and data["entity_id"] == -1:
+            if data["entity_type"] == "Project":
+                data["entity_id"] = data["project_id"]
+            else:
+                temp_entity = self._engine.shotgun.find_one(
+                    data["entity_type"],
+                    [["project", "is", dict(type="Project", id=data["project_id"])]],
+                )
+                data["entity_id"] = temp_entity["id"]
+
+        project_entity, entities = self._get_entities_from_payload(data)
+        entity = entities[0]
+
+        # We can end up getting requests for actions for any entity type
+        # that exists in Shotgun. Many of those we don't want to provide
+        # commands for, so we can just notify the client that it's an
+        # unsupported entity type.
+        env_name = self.__pick_environment(project_entity, entity)
+
+        if env_name is None:
+            self.host.reply(dict(retcode=self.UNSUPPORTED_ENTITY_TYPE))
+            return
+
+        manager = sgtk.bootstrap.ToolkitManager()
+        manager.base_configuration = self.BASE_CONFIG_URI
+
+        pcs = manager.sort_and_filter_configuration_entities(
+            project=project_entity,
+            entities=data["pipeline_configs"],
```

Comment:
The thought here was that because the action menu in SG already has the PC entities queried, that we could reuse that data instead of re-querying it. That being said, now that I've shifted over to querying things from Python (like Software entities, as an example), it might make more sense to do the same for PC entities.